src/main/java/io/cdap/plugin/marketo/common/api/MarketoHttp.java

```diff
@@ -0,0 +1,204 @@
+    return new MarketoPageIterator<>(getPage(queryUrl, pageClass), this, queryUrl, pageClass, resultsGetter);
+  }
+
+  <T extends BaseResponse> T validatedGet(String queryUrl, Map<String, String> parameters,
+                                          Function<InputStream, T> deserializer) {
+    String logUri = "GET " + buildUri(queryUrl, parameters, false).toString();
+    return retryableValidate(logUri, () -> {
+      URI queryUri = buildUri(queryUrl, parameters, true);
+      return get(queryUri, deserializer);
+    });
+  }
+
+  public <T extends BaseResponse, B> T validatedPost(String queryUrl, Map<String, String> parameters,
+                                              Function<InputStream, T> deserializer,
+                                              B body, Function<B, String> qSerializer) {
+    String logUri = "POST " + buildUri(queryUrl, parameters, false).toString();
+    return retryableValidate(logUri, () -> {
+      URI queryUri = buildUri(queryUrl, parameters, true);
+      return post(queryUri, deserializer, body, qSerializer);
+    });
+  }
+
+  private <T extends BaseResponse> T retryableValidate(String logUri, Supplier<T> tryQuery) {
+    T result = tryQuery.get();
+    // check for expired token
+    if (!result.isSuccess()) {
+      for (Error error : result.getErrors()) {
+        if (error.getCode() == 602 && error.getMessage().equals("Access token expired")) {
+          // refresh token and retry
+          token = refreshToken();
+          LOG.info("Refreshed token");
+          return tryQuery.get();
+        }
+      }
+    }
+
+    // log warnings if required
+    if (result.getWarnings().size() > 0) {
+      String warnings = result.getWarnings().stream()
+        .map(error -> String.format("code: %s, message: %s", error.getCode(), error.getMessage()))
+        .collect(Collectors.joining("; "));
+      LOG.warn("Warnings when calling '{}' - {}", logUri, warnings);
+    }
+
+    if (!result.isSuccess()) {
+      String msg = String.format("Errors when calling '%s'", logUri);
+      // log errors if required
+      if (result.getErrors().size() > 0) {
+        String errors = result.getErrors().stream()
+          .map(error -> String.format("code: %s, message: %s", error.getCode(), error.getMessage()))
+          .collect(Collectors.joining("; "));
+        msg = msg + " - " + errors;
+        LOG.error(msg);
+      }
+      throw new RuntimeException(msg);
+    }
+    return result;
+  }
+
+  public <T> T get(URI uri, Function<InputStream, T> deserializer) {
+    try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
+      HttpGet request = new HttpGet(uri);
+      try (CloseableHttpResponse response = httpClient.execute(request, httpClientContext)) {
+        if (response.getStatusLine().getStatusCode() >= 300) {
+          throw new IOException(IOUtils.toString(response.getEntity().getContent(), StandardCharsets.UTF_8));
```

Comment:
What about some message, instead of just plain text. Like "Request to '%s' failed with '%d' code and '%s' output."

Reply:
Helper method under given 'try' will wrap all exceptions to descriptive message.
Http code was missing, adding that.